trin-core/src/utp/packets.rs

```diff
@@ -0,0 +1,322 @@
+            return Err("Invalid packet length");
+        }
+        extension = data[index];
+        let len = data[index + 1] as usize;
+
+        if len == 0 || len % 4 != 0 || len + index + 2 > data.len() {
+            return Err("Invalid Extension Length");
+        }
+
+        index += len + 2;
+    }
+    Ok(())
+}
+
+impl<'a> TryFrom<&'a [u8]> for Packet {
+    type Error = &'static str;
+
+    // todo: Added support to check validity of extensions,
+    // we don't need to worry about checking payloads
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        match PacketHeader::try_from(value).and(check_extensions(value)) {
+            Ok(_) => Ok(Packet(value.to_owned())),
+            Err(e) => Err(e),
+        }
+    }
+}
+
+impl Default for PacketHeader {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub struct PacketHeader {
+    type_ver: u8,
+    extension: u8,
+    connection_id: u16,
+    // This time is in microseconds
+    timestamp: u32,
+    timestamp_difference: u32,
+    wnd_size: u32,
+    pub seq_nr: u16,
+    ack_nr: u16,
+}
+
+impl PacketHeader {
+    pub fn new() -> Self {
+        PacketHeader {
+            type_ver: u8::from(PacketType::Data) << 4 | VERSION,
+            extension: 0,
+            connection_id: 0,
+            timestamp: 0,
+            timestamp_difference: 0,
+            wnd_size: 0xf000,
+            seq_nr: 0,
+            ack_nr: 0,
+        }
+    }
+
+    pub fn encode(&mut self) -> Vec<u8> {
+        let mut buf = vec![];
+        buf.extend_from_slice(&[self.type_ver]);
+        buf.extend_from_slice(&[self.extension]);
+        buf.extend_from_slice(&self.connection_id.to_be_bytes());
+        buf.extend_from_slice(&self.timestamp.to_be_bytes());
+        buf.extend_from_slice(&self.timestamp_difference.to_be_bytes());
+        buf.extend_from_slice(&self.wnd_size.to_be_bytes());
+        buf.extend_from_slice(&self.seq_nr.to_be_bytes());
+        buf.extend_from_slice(&self.ack_nr.to_be_bytes());
+        buf
+    }
+
+    pub fn decode(bytes: &[u8]) -> PacketHeader {
+        PacketHeader {
+            type_ver: bytes[0],
+            extension: bytes[1],
+            connection_id: u16::from_be_bytes([bytes[2], bytes[3]]),
+            timestamp: u32::from_be_bytes([bytes[4], bytes[5], bytes[6], bytes[7]]),
+            timestamp_difference: u32::from_be_bytes([bytes[8], bytes[9], bytes[10], bytes[11]]),
+            wnd_size: u32::from_be_bytes([bytes[12], bytes[13], bytes[14], bytes[15]]),
+            seq_nr: u16::from_be_bytes([bytes[16], bytes[17]]),
+            ack_nr: u16::from_be_bytes([bytes[18], bytes[19]]),
+        }
+    }
+
+    pub fn version(&self) -> u8 {
+        self.type_ver & 0xf
+    }
+    pub fn set_version(&mut self, int: u8) {
+        self.type_ver = (self.type_ver & 0xf0) | (int & 0xf)
+    }
+    pub fn type_(&self) -> PacketType {
+        PacketType::try_from(self.type_ver >> 4).unwrap()
```

Comment:
Should we handle this error instead of unwrapping it?

Reply:
This will never report an error, it is deserialized and checked if there is an issue when we originally decode the incoming packet, we have to do try_from just because of Rust unless we want to use unsafe to do this.

Reply:
What do you mean this can't ever fail? The `TryFrom` trait always return a `Result<T, E>` and we implement it here:

```rust
impl TryFrom<u8> for PacketType {
    type Error = &'static str;

    fn try_from(value: u8) -> Result<Self, Self::Error> {
        match value {
            0 => Ok(PacketType::Data),
            1 => Ok(PacketType::Fin),
            2 => Ok(PacketType::State),
            3 => Ok(PacketType::Reset),
            4 => Ok(PacketType::Syn),
            _ => Err("invalid packet type"),
        }
    }
}
```

So if we pass a `value = 5` it will fail. There is a `From` trait that doesn't return an error. I understand that in the current implementation we may have additional assertions to never pass invalid value but in a future refactoring, for example, this may not be the case and we want our code to be foolproof especially when we are building a utp library.

Reply:
![image](https://user-images.githubusercontent.com/31669092/150179554-98d7ea81-e4bc-4f62-8fb9-db4bfc6cbb9d.png)

We check if it is invalid when we first receive the packet from a 3rd party client before we even handle it. As seen in the picture. Internally we can't set the value to 5, because it can only be set from the enum PacketType.